static/js/chart/draw_bar.ts

```diff
@@ -804,41 +806,86 @@ export function drawHorizontalBarChart(
     : Math.ceil((dataGroups.length + 0.1) * barHeight * numGroups) +
       marginTop +
       marginBottom;
-
-  const x = d3
-    .scaleLinear()
-    .domain([
-      0,
-      options?.stacked
-        ? d3.max(series, (d) => d3.max(d, (d) => d[1]))
-        : d3.max(dataGroups, (dg) => d3.max(dg.value, (dgv) => dgv.value)),
-    ])
-    .nice()
-    .rangeRound([marginLeft, chartWidth - marginRight]);
-
-  const y = d3
-    .scaleBand()
-    .domain(dataGroups.map((dg) => dg.label))
-    .domain(dataGroups.map((dg) => dg.label))
-    .rangeRound([marginTop, height - marginBottom])
-    .padding(0.15);
-
   const color = getColorFn(keys, options?.colors);
+  const [displayUnit, label] = getDisplayUnitAndLabel(options?.unit);
 
   // Create the SVG container.
   const svg = d3
     .create("svg")
+    .attr("class", ASYNC_ELEMENT_CLASS) // for render verification in test cases
     .attr("width", chartWidth)
     .attr("height", height)
     .attr("viewBox", `0, 0, ${chartWidth}, ${height}`)
     .attr("style", "max-width: 100%; height: auto; font: 10px sans-serif;");
 
-  if (options?.stacked) {
-    // Stacked bar chart
-    drawHorizontalStackedBars(svg, color, series, x, y, options?.lollipop);
-  } else {
-    drawHorizontalGroupedBars(svg, color, dataGroups, x, y, options?.lollipop);
+  // Attach SVG node to the parent container
+  // The node must be attached before any getBBox() calls.
+  containerElement.append(svg.node());
+
+  if (label) {
+    // Add x-axis label in the top-middle of the axis
+    svg
+      .append("text")
+      .attr("id", "x-axis-label")
+      .attr("part", "x-axis-label")
+      .attr("x", chartWidth / 2)
+      .attr("y", 0)
+      .attr("fill", AXIS_TEXT_FILL)
+      .attr("text-anchor", "middle")
+      .attr("dominant-baseline", "hanging")
+      .text(label);
+
+    // Adjust top margin to account for any x-axis labels
+    const xAxisLabelHeight = (
+      svg.select("#x-axis-label").node() as SVGGraphicsElement
+    ).getBBox().height;
+    marginTop += xAxisLabelHeight;
+  }
+
+  const y = d3
+    .scaleBand()
+    .domain(dataGroups.map((dg) => dg.label))
+    .domain(dataGroups.map((dg) => dg.label))
+    .rangeRound([marginTop, height - marginBottom])
+    .padding(0.15);
+
+  // y axis
+  const yAxis = svg
+    .append("g")
+    .call(d3.axisLeft(y).tickSizeOuter(0))
+    .call((g) =>
+      g
+        .selectAll("text")
+        .style("fill", AXIS_TEXT_FILL)
+        .style("shape-rendering", "crispEdges")
+    );
+
+  // Get maximum length of tick labels
+  let maxLabelWidth = marginLeft;
+  if (!options.style.yAxisMargin) {
```

Comment:
if yAxisMargin is set, does that override automatically determining the label width?